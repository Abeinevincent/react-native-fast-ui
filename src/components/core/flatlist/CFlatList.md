# CFlatList Component

A custom `FlatList` component with additional styling capabilities to enhance faster styling.

## Props

### `contentContainerStyle`

- Type: `ViewStyle`
- Description: Style object for the content container of the FlatList.

### `w`

- Type: `number`
- Default: `screenWidth * 0.9`
- Description: Width of the FlatList. If not provided, it defaults to 90% of the screen width.

### `h`

- Type: `number`
- Default: `screenHeight * 0.9`
- Description: Height of the FlatList. If not provided, it defaults to 90% of the screen height.

### `bgc`

- Type: `string`
- Default: `'red'`
- Description: Background color of the FlatList. If not provided, it defaults to red.

### `...rest`

- Type: `ICFlatList<T>`
- Description: Additional props passed to the FlatList component. These props should conform to the `ICFlatList` type.

### [CView Props](../view/CView.md)

## Example Usage

```jsx
import React from 'react';
import { CFlatList } from './path/to/CFlatList';
import { StyleSheet } from 'react-native';

const data = [...]; // Your data array

const renderItem = ({ item }) => (
  // Render your item here
);

const MyComponent = () => {
  return (
    <CFlatList
      data={data}
      renderItem={renderItem}
      contentContainerStyle={styles.container}
      w={300}
      h={400}
      bgc="blue"
    />
  );
};

const styles = StyleSheet.create({
  container: {
    // Your custom content container styles
  },
});

export default MyComponent;
```
