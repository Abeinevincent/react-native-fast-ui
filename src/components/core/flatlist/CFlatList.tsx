import React from 'react';
import { FlatList, StyleSheet, type ViewStyle } from 'react-native';
import type { ICFlatList } from './CFlatList.type';
import { viewStyleMapper } from '../../utils/stylemapper';
import type { ICView } from '../view/CView.types';

const CFlatList = <T,>({
  contentContainerStyle,
  bgc,
  ...rest
}: ICFlatList<T>) => {
  // Create default style object
  const flatListStyle: ViewStyle = {
    backgroundColor: bgc ? bgc : 'white',
  };

  // Iterate through props and apply corresponding style props dynamically
  Object.keys(rest).forEach((prop) => {
    const key = prop as keyof ICView;
    if (viewStyleMapper()[key]) {
      // @ts-ignore
      flatListStyle[viewStyleMapper()[key]] = rest[key];
    }
  });

  // Combine custom styles with additional styles
  const combinedStyles = StyleSheet.flatten([
    flatListStyle,
    contentContainerStyle,
  ]);

  // Render FlatList component with combined styles and other props
  return (
    <FlatList
      testID="CFlatList"
      {...rest}
      contentContainerStyle={combinedStyles}
    />
  );
};

export { CFlatList };
