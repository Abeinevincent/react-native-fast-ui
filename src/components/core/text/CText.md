# CText Component

The `CText` component is a custom text component that extends the functionality of the React Native `Text` component.

## Props

### Children Props

- `children`: (Required) Content to render on the ui.

### Style-related Props

- `w`: (Optional) Width of the text. Accepts a string or a number. Default: '100%'.
- `h`: (Optional) Height of the text. Accepts a string or a number. Default: 'auto'.
- `c`: (Optional) Text color. Accepts a string representing a color code. Default: 'black'.
- `ta`: (Optional) Text alignment. Accepts one of: 'auto', 'left', 'right', 'center', 'justify'. Default: 'auto'.
- `tav`: (Optional) Vertical text alignment. Accepts one of: 'auto', 'top', 'bottom', 'center'. Default: 'auto'.
- `ff`: (Optional) Font family. Accepts a string representing the font family. Default: 'System'.
- `fs`: (Optional) Font size. Accepts a number representing font size. Default: 14.
- `fst`: (Optional) Font style. Accepts one of: 'normal', 'italic'. Default: 'normal'.
- `fw`: (Optional) Font weight. Accepts one of: 'normal', 'bold', or a numeric value from '100' to '900'. Default: 'normal'.
- `ls`: (Optional) Letter spacing. Accepts a number representing letter spacing. Default: 0.
- `lh`: (Optional) Line height. Accepts a number representing line height. Default: 20.
- `tdl`: (Optional) Text decoration line. Accepts one of: 'none', 'underline', 'line-through', 'underline line-through'. Default: 'none'.
- `tds`: (Optional) Text decoration style. Accepts one of: 'solid', 'double', 'dotted', 'dashed'. Default: 'solid'.
- `tdc`: (Optional) Text decoration color. Accepts a string representing a color code. Default: 'black'.
- `tsc`: (Optional) Text shadow color. Accepts a string representing a color code. Default: 'transparent'.
- `tso`: (Optional) Text shadow offset. Accepts an object with `width` and `height` properties. Default: `{ width: 0, height: 0 }`.
- `tsr`: (Optional) Text shadow radius. Accepts a number representing the text shadow radius. Default: 0.
- `tt`: (Optional) Text transform. Accepts one of: 'none', 'uppercase', 'lowercase', 'capitalize'. Default: 'none'.
- `bber`: (Optional) Border bottom end radius. Accepts a number.
- `bblr`: (Optional) Border bottom left radius. Accepts a number.
- `bbrr`: (Optional) Border bottom right radius. Accepts a number.
- `bbsr`: (Optional) Border bottom start radius. Accepts a number.
- `bser`: (Optional) Border start end radius. Accepts a number.
- `bssr`: (Optional) Border start start radius. Accepts a number.
- `beer`: (Optional) Border end end radius. Accepts a number.
- `besr`: (Optional) Border end start radius. Accepts a number.
- `bbw`: (Optional) Border bottom width. Accepts a number.
- `bc`: (Optional) Border color. Accepts a string representing a color code.
- `bci`: (Optional) Border curve iOS. Accepts one of: 'circular', 'continuous'.
- `bec`: (Optional) Border end color. Accepts a string representing a color code.
- `blc`: (Optional) Border left color. Accepts a string representing a color code.
- `blw`: (Optional) Border left width. Accepts a number.
- `br`: (Optional) Border radius. Accepts a number.
- `brc`: (Optional) Border right color. Accepts a string representing a color code.
- `brw`: (Optional) Border right width. Accepts a number.
- `bsc`: (Optional) Border start color. Accepts a string representing a color code.
- `bs`: (Optional) Border style. Accepts one of: 'solid', 'dotted', 'dashed'.
- `btc`: (Optional) Border top color. Accepts a string representing a color code.
- `bter`: (Optional) Border top end radius. Accepts a number.
- `btlr`: (Optional) Border top left radius. Accepts a number.
- `btrr`: (Optional) Border top right radius. Accepts a number.
- `btsr`: (Optional) Border top start radius. Accepts a number.
- `btw`: (Optional) Border top width. Accepts a number.
- `bw`: (Optional) Border width. Accepts a number.
- `el`: (Optional) Elevation Android. Sets the elevation of the view. Accepts a number.
- `o`: (Optional) Opacity. Accepts a number.

### Margin and Padding Props

- `m`: (Optional) Margin. Accepts a number or a string representing margin value. Default: 0.
- `mt`: (Optional) Margin top. Accepts a number or a string representing margin top value. Default: 0.
- `mr`: (Optional) Margin right. Accepts a number or a string representing margin right value. Default: 0.
- `mb`: (Optional) Margin bottom. Accepts a number or a string representing margin bottom value. Default: 0.
- `ml`: (Optional) Margin left. Accepts a number or a string representing margin left value. Default: 0.
- `p`: (Optional) Padding. Accepts a number or a string representing padding value. Default: 0.
- `pt`: (Optional) Padding top. Accepts a number or a string representing padding top value. Default: 0.
- `pr`: (Optional) Padding right. Accepts a number or a string representing padding right value. Default: 0.
- `pb`: (Optional) Padding bottom. Accepts a number or a string representing padding bottom value. Default: 0.
- `pl`: (Optional) Padding left. Accepts a number or a string representing padding left value. Default: 0.

### Platform-specific Props

- `ifp`: (Optional) Include font padding (Android). Accepts a boolean. Default: true.
- `va`: (Optional) Vertical align (Android). Accepts one of: 'auto', 'top', 'bottom', 'middle'. Default: 'auto'.
- `wd`: (Optional) Writing direction (iOS). Accepts one of: 'auto', 'ltr', 'rtl'. Default: 'auto'.

### Universal Props

- `us`: (Optional) User select. Accepts one of: 'auto', 'text', 'none', 'contain', 'all'. Default: 'none'.

## Example Usage

```jsx
import { CText } from './CText';

// Example usage of CText component
<CText
  w={100}
  h={50}
  c="red"
  fontSize={20}
  fontWeight="bold"
  margin={10}
  padding={5}
>
  Custom Text
</CText>;
```
