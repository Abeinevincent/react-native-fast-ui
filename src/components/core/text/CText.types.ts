import {
  type ColorValue,
  type DimensionValue,
  type TextProps,
} from 'react-native';
import { type IReusableStyleTypes } from '../../../types/reusabletypes/ReusableStyleTypes';
import type { ReactNode } from 'react';

// Define interface for CText props extending TextProps
export interface ICText extends TextProps, IReusableStyleTypes {
  // Children
  children: ReactNode;
  // Style-related props
  c?: string; // Text color prop
  bgc?: ColorValue | string; // backgroundColor
  h?: DimensionValue | undefined;
  w?: DimensionValue | undefined;
  ta?: 'auto' | 'left' | 'right' | 'center' | 'justify'; // text align
  tav?: 'auto' | 'top' | 'bottom' | 'center'; //text align vertical
  ff?: string; // Font family prop
  fs?: number; // Font size prop
  fst?: 'normal' | 'italic'; // Font style prop
  fw?:
    | 'normal'
    | 'bold'
    | '100'
    | '200'
    | '300'
    | '400'
    | '500'
    | '600'
    | '700'
    | '800'
    | '900'; // Font weight prop
  ls?: number; // Letter spacing prop
  lh?: number; // Line height prop
  tdl?: 'none' | 'underline' | 'line-through' | 'underline line-through'; // Text decoration line prop
  tds?: 'solid' | 'double' | 'dotted' | 'dashed'; // Text decoration style prop
  tdc?: string; // Text decoration color prop
  tsc?: string; // Text shadow color prop
  tso?: { width: number; height: number }; // Text shadow offset prop
  tsr?: number; // Text shadow radius prop
  tt?: 'none' | 'uppercase' | 'lowercase' | 'capitalize'; // Text transform prop
  // Android-specific props
  ifp?: boolean; // Include font padding prop
  va?: 'auto' | 'top' | 'bottom' | 'middle'; // Vertical align prop
  // iOS-specific props
  wd?: 'auto' | 'ltr' | 'rtl'; // Writing direction prop
  // Universal props
  us?: 'auto' | 'text' | 'none' | 'contain' | 'all'; // User select prop
}
