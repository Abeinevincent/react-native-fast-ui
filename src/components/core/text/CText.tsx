import React from 'react';
import { Text, Platform, type TextStyle } from 'react-native';
import { type ICText } from './CText.types';
import { textStyleMapper } from '../../utils/stylemapper';

const CText: React.FC<ICText> = ({
  style: additionalStyles,
  children,
  fs,
  c,
  ...rest
}) => {
  // Create style object for default styles
  const textStyle: TextStyle = {
    fontSize: fs ? fs : 16,
    color: c ? c : 'black',
  };

  // Iterate through props and apply corresponding style props dynamically
  Object.keys(rest).forEach((prop) => {
    const key = prop as keyof ICText;
    if (textStyleMapper()[key]) {
      // @ts-ignore
      textStyle[textStyleMapper()[key]] = rest[key];
    }
  });

  // Apply platform-specific styles
  if (Platform.OS === 'android') {
    textStyle.includeFontPadding = rest.ifp ?? true;
    textStyle.verticalAlign = rest.va ?? 'auto';
  } else if (Platform.OS === 'ios') {
    textStyle.writingDirection = rest.wd ?? 'auto';
  }

  // Combine custom styles with additional styles
  const combinedStyles = [textStyle, additionalStyles];

  // Render CText component with combined styles and children
  return (
    <Text style={combinedStyles} {...rest} testID="text">
      {children}
    </Text>
  );
};

export { CText };
