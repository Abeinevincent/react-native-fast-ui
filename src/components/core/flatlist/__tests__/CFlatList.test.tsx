import React from 'react';
import { render } from '@testing-library/react-native';
import '@testing-library/jest-native/extend-expect'; // Import this to extend Jest's expect functionality
import { CText } from '../../text/CText';
import { CFlatList } from '../CFlatList';

describe('CFlatList component', () => {
  it('renders correctly with provided data', () => {
    const DATA = [
      { id: '1', title: 'First Item' },
      { id: '2', title: 'Second Item' },
      { id: '3', title: 'Third Item' },
    ];

    const renderItem = ({
      item,
    }: {
      item: {
        id: string;
        title: string;
      };
    }) => <CText>{item.title}</CText>;
    const { getByTestId, toJSON } = render(
      <CFlatList
        data={DATA}
        testID="CFlatList"
        renderItem={renderItem}
        keyExtractor={(item: { id: string; title: string }) =>
          item.id.toString()
        }
      />
    );
    const flatList = getByTestId('CFlatList');
    expect(flatList).toBeTruthy();
    expect(toJSON()).toMatchSnapshot();
  });

  it('renders items correctly', () => {
    const DATA = [
      { id: '1', title: 'First Item' },
      { id: '2', title: 'Second Item' },
      { id: '3', title: 'Third Item' },
    ];

    const renderItem = ({
      item,
    }: {
      item: {
        id: string;
        title: string;
      };
    }) => <CText>{item.title}</CText>;
    const { getByText } = render(
      <CFlatList
        data={DATA}
        testID="CFlatList"
        renderItem={renderItem}
        keyExtractor={(item) => item.id.toString()}
      />
    );
    expect(getByText('First Item')).toBeTruthy();
    expect(getByText('Second Item')).toBeTruthy();
    expect(getByText('Third Item')).toBeTruthy();
  });

  // Add more test cases for styles, props, etc. as needed
});
