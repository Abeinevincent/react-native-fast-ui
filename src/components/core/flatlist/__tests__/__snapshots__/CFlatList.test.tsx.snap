// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CFlatList component renders correctly with provided data 1`] = `
<RCTScrollView
  contentContainerStyle={
    {
      "backgroundColor": "white",
    }
  }
  data={
    [
      {
        "id": "1",
        "title": "First Item",
      },
      {
        "id": "2",
        "title": "Second Item",
      },
      {
        "id": "3",
        "title": "Third Item",
      },
    ]
  }
  getItem={[Function]}
  getItemCount={[Function]}
  keyExtractor={[Function]}
  onContentSizeChange={[Function]}
  onLayout={[Function]}
  onMomentumScrollBegin={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={0.0001}
  stickyHeaderIndices={[]}
  testID="CFlatList"
  viewabilityConfigCallbackPairs={[]}
>
  <View>
    <View
      onFocusCapture={[Function]}
      onLayout={[Function]}
      style={null}
    >
      <Text
        style={
          [
            {
              "color": "black",
              "fontSize": 16,
              "writingDirection": "auto",
            },
            undefined,
          ]
        }
        testID="text"
      >
        First Item
      </Text>
    </View>
    <View
      onFocusCapture={[Function]}
      onLayout={[Function]}
      style={null}
    >
      <Text
        style={
          [
            {
              "color": "black",
              "fontSize": 16,
              "writingDirection": "auto",
            },
            undefined,
          ]
        }
        testID="text"
      >
        Second Item
      </Text>
    </View>
    <View
      onFocusCapture={[Function]}
      onLayout={[Function]}
      style={null}
    >
      <Text
        style={
          [
            {
              "color": "black",
              "fontSize": 16,
              "writingDirection": "auto",
            },
            undefined,
          ]
        }
        testID="text"
      >
        Third Item
      </Text>
    </View>
  </View>
</RCTScrollView>
`;
